plugin/cshared.go

```diff
@@ -0,0 +1,279 @@
+	defer initWG.Done()
+
+	registerWG.Wait()
+
+	if theInput == nil && theOutput == nil {
+		fmt.Fprintf(os.Stderr, "no input or output registered\n")
+		return input.FLB_RETRY
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var err error
+	if theInput != nil {
+		conf := &flbInputConfigLoader{ptr: ptr}
+		err = theInput.Init(ctx, conf)
+	} else {
+		conf := &flbOutputConfigLoader{ptr: ptr}
+		err = theOutput.Init(ctx, conf)
+	}
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "init: %v\n", err)
+		return input.FLB_ERROR
+	}
+
+	return input.FLB_OK
+}
+
+//export FLBPluginInputCallback
+func FLBPluginInputCallback(data *unsafe.Pointer, csize *C.size_t) int {
+	initWG.Wait()
+
+	if theInput == nil {
+		fmt.Fprintf(os.Stderr, "no input registered\n")
+		return input.FLB_RETRY
+	}
+
+	var err error
+	once.Do(func() {
+		runCtx, runCancel = context.WithCancel(context.Background())
+		theChannel = make(chan Message, 1)
+		go func() {
+			err = theInput.Collect(runCtx, theChannel)
+		}()
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "run: %s\n", err)
+		return input.FLB_ERROR
+	}
+
+	select {
+	case msg, ok := <-theChannel:
+		if !ok {
+			return input.FLB_OK
+		}
+
+		t := input.FLBTime{Time: msg.Time}
+		b, err := input.NewEncoder().Encode([]any{t, msg.Record})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "encode: %s\n", err)
+			return input.FLB_ERROR
+		}
+
+		cdata := C.CBytes(b)
+
+		*data = cdata
+		*csize = C.size_t(len(b))
+
+		// C.free(unsafe.Pointer(cdata))
+	case <-runCtx.Done():
+		err := runCtx.Err()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			fmt.Fprintf(os.Stderr, "run: %s\n", err)
+			return input.FLB_ERROR
+		}
+	}
+
+	return input.FLB_OK
+}
+
+//export FLBPluginFlush
+func FLBPluginFlush(data unsafe.Pointer, clength C.int, ctag *C.char) int {
+	initWG.Wait()
+
+	if theOutput == nil {
+		fmt.Fprintf(os.Stderr, "no output registered\n")
+		return output.FLB_RETRY
+	}
+
+	var err error
+	once.Do(func() {
+		runCtx, runCancel = context.WithCancel(context.Background())
+		theChannel = make(chan Message, 1)
+		go func() {
+			err = theOutput.Flush(runCtx, theChannel)
+		}()
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "run: %s\n", err)
+		return output.FLB_ERROR
+	}
+
+	select {
+	case <-runCtx.Done():
+		err := runCtx.Err()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			fmt.Fprintf(os.Stderr, "run: %s\n", err)
+			return output.FLB_ERROR
+		}
+
+		return output.FLB_OK
+	default:
+	}
+
+	in := C.GoBytes(data, C.int(clength))
+	h := &codec.MsgpackHandle{}
+	h.SetBytesExt(reflect.TypeOf(fTime{}), 0, &fTime{})
+	dec := codec.NewDecoderBytes(in, h)
+
+	for {
+		select {
+		case <-runCtx.Done():
+			err := runCtx.Err()
+			if err != nil && !errors.Is(err, context.Canceled) {
+				fmt.Fprintf(os.Stderr, "run: %s\n", err)
+				return output.FLB_ERROR
+			}
+
+			return output.FLB_OK
+		default:
+		}
+
+		var entry []any
+		err := dec.Decode(&entry)
```

Comment:
I don't know if all these assertions of `[]any{fTime, map[string]string}` could potentially broke something.

I saw the first input plugin implementation also checked for `int64` besides `fTime`.
But I guess that was because the msgpack writer was exposed directly and the plugin author could have been writen anything really. ANd it was counting for the case the plugin was sending it as unix seconds.

Reply:
No, internally it won't have a different format. I think your assumption is correct. []any{fTime, map[string]string}